### PR TITLE
Implement metrics reporting via opentelemetry

### PR DESCRIPTION
**Pull request by Ardagan**
**What changed?**
* Add generalized Prometheus config
* Refactor metrics initialization code to utilize generalized
Prometheus config
* Add OpenTelemetry Client, Scope and Reporter implementations

**Why?**
Add support for maintained OpenTelemetry in addition to stale Tally library.

**How did you test it?**
Manual testing, automated tests.

**Potential risks**
Metrics are not reported.
Metric tags are not injected from config.

**Is hotfix candidate?**
No

### CODE REVIEW

File: common/metrics/common.go [hunk 1]
```diff
@@ -48,3 +54,22 @@ func getMetricDefs(serviceIdx ServiceIdx) map[int]metricDefinition {
...
+	case primitives.MatchingService:
+		return Matching
+	case primitives.WorkerService:
+		return Worker
+	default:
+		logger.Fatal("Unknown service name for metrics!", tag.Service(serviceName))
+	}
+
+	// this should never happen!
+	return NumServices
```
Comment (wxing1292) [inline common/metrics/common.go:74]: why not just panic in the default?

File: common/metrics/opentelemetery_reporter.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		exporter  *prometheus.Exporter
+		meter     metric.Meter
+		meterMust metric.MeterMust
+		tags      map[string]string
+		prefix    string
+		config    *PrometheusConfig
+	}
+)
+
+func newOpentelemeteryReporter(logger log.Logger, tags map[string]string, prefix string, prometheusConfig *PrometheusConfig) (*OpentelemetryReporter, error) {
```
Comment (wxing1292) [inline common/metrics/opentelemetery_reporter.go:50]: format?

File: common/metrics/opentelemetery_reporter.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+func initPrometheusListener(config *PrometheusConfig, logger log.Logger, exporter *prometheus.Exporter) {
+	handlerPath := config.HandlerPath
+	if handlerPath == "" {
+		handlerPath = "/metrics"
+	}
+	http.HandleFunc(handlerPath, exporter.ServeHTTP)
+	go func() {
+		logger.Info("Starting prometheus listener.", tag.Address(config.ListenAddress))
+		err := http.ListenAndServe(config.ListenAddress, nil)
```
Comment (wxing1292) [inline common/metrics/opentelemetery_reporter.go:89]: how does this handle shutdown?

File: common/metrics/config.go [hunk 4]
```diff
@@ -131,9 +148,70 @@ var (
...
 		500000 * ms,
 		1000000 * ms,
-	})
+	}
 )
 
+// InitMetricReporters is a root function for initalizing metrics clients.
+//
+// Usage pattern
+// serverReporter, sdkReporter, err := c.InitMetricReporters
```
Comment (alexshtin) [inline common/metrics/config.go:157]:
```suggestion
// serverReporter, sdkReporter, err := c.InitMetricReporters(logger, xxx)
// where xxx should be of type ...
```

File: common/metrics/config.go [hunk 5]
```diff
@@ -131,9 +148,70 @@ var (
...
+// metricsClient := serverReporter.newClient(logger, serviceIdx)
+//
+// returns SeverReporter, SDKReporter, error
+func (c *Config) InitMetricReporters(logger log.Logger, extensionPoint interface{}) (Reporter, Reporter, error) {
```
Comment (alexshtin) [inline common/metrics/config.go:161]: I would rename `extensionPoint` to `customReporter`.

File: common/metrics/config.go [hunk 6]
```diff
@@ -131,9 +148,70 @@ var (
...
+	if c.PrometheusServer == nil {
+		var scope tally.Scope
+		if extensionPoint != nil {
```
Comment (alexshtin) [inline common/metrics/config.go:164]:
```suggestion
		if tallyCustomReporter, ok := extensionPoint.(tally.BaseStatsReporter); ok {
```

File: common/metrics/common.go (hunk 1 above)
Comment (Ardagan) [reply common/metrics/common.go:74]: logger.Fatal already does panic. I'll add explicit panic as well.

File: common/metrics/noop_metrics_client.go [hunk 7]
```diff
@@ -53,7 +53,7 @@ func (m NoopMetricsClient) RecordDistribution(scope int, timer int, d int) {}
 func (m NoopMetricsClient) UpdateGauge(scope int, gauge int, value float64) {}
 
 func (m NoopMetricsClient) Scope(scope int, tags ...Tag) Scope {
-	return &metricsScope{
+	return &tallyScope{
```
[the remark this answers is not in the document]
Comment (Ardagan) [reply common/metrics/noop_metrics_client.go:56]: That's a good comment, but I think it's best to do in separate PR.

File: common/metrics/common.go [hunk 8]
```diff
@@ -48,3 +55,20 @@ func getMetricDefs(serviceIdx ServiceIdx) map[int]metricDefinition {
...
+	case primitives.FrontendService:
+		return Frontend
+	case primitives.HistoryService:
+		return History
+	case primitives.MatchingService:
+		return Matching
+	case primitives.WorkerService:
+		return Worker
+	default:
+		logger.Fatal("Unknown service name for metrics!", tag.Service(serviceName))
```
Comment (wxing1292) [inline common/metrics/common.go:71]: i mean you can remove the log.Fatal and let the panic do the work

File: config/development.yaml [hunk 9]
```diff
@@ -19,8 +19,14 @@ global:
     port: 7936
   metrics:
     prometheus:
+#      framework: "opentelemetry"
+#      framework: "tally"
```
Comment (wxing1292) [inline config/development.yaml:23]: remove or make this comment more easier to understand?

**Review by wxing1292 — APPROVED** [review]

**Pull request merged**